### PR TITLE
Extend reshape hack

### PR DESCRIPTION
**Pull request by madt2709**
Avoid merge views in certain symbolic cases as well.

This test:

```python
  def test_expand_into_reshape_with_symbols(self):
    vi = Variable("i", 1, 5)
    a = Tensor([[1], [2], [3]]).expand((3, 4))
    a = a.reshape(3, vi)
    assert a.shape == (3, vi)
    assert a.lazydata.var_vals == {vi: 4}
```
currently fails on master due to a type error. Need to think if thats a bug or intentional

### CODE REVIEW

File: tinygrad/shape/view.py [hunk 1]
```diff
@@ -97,7 +98,7 @@ def reshape(self, new_shape: Tuple[sint, ...]) -> Optional[View]:
 
     # check if this is adding or removing 1s (only)
     # NOTE: this is optional, but removes most calls to (expensive!) merge_views (with mask, not optional)
-    if [x for x in self.shape if x != 1] == [x for x in new_shape if x != 1]:
+    if sum(1 for a,b in zip_longest([x for x in self.shape if x != 1],[x for x in new_shape if x != 1]) if a != b) <= 1:
```
Comment (geohot) [inline tinygrad/shape/view.py:101]: I don't understand this, what does this catch that the old one didn't?

Comment (madt2709) [reply tinygrad/shape/view.py:101]: This catches reshapes like (3,4) --> (3, i) where an int gets replaced with a Variable and the other way round.

Comment (madt2709) [reply tinygrad/shape/view.py:102]: Added a comment for this.

Comment (geohot) [reply tinygrad/shape/view.py:102]: oh I would handle this explicitly then

Comment (madt2709) [reply tinygrad/shape/view.py]: updated it to (hopefully!) be more readable

File: tinygrad/shape/view.py [hunk 2]
```diff
@@ -95,9 +96,12 @@ def reshape(self, new_shape: Tuple[sint, ...]) -> Optional[View]:
     # after the asserts, it's okay to check contiguous
     if self.contiguous: return View.create(new_shape)
 
-    # check if this is adding or removing 1s (only)
+    # check if this is adding or removing 1s with [x for x in self.shape if x != 1] == [x for x in new_shape if x != 1].
+    # extend the above to include cases like (3,4) --> (3,i) where we add or remove 1s and add/remove a "dummy" variable.
     # NOTE: this is optional, but removes most calls to (expensive!) merge_views (with mask, not optional)
-    if [x for x in self.shape if x != 1] == [x for x in new_shape if x != 1]:
+    diff_between_shapes = [(a,b) for a,b in zip_longest([x for x in self.shape if x != 1],[x for x in new_shape if x != 1]) if a != b]
```
Comment (geohot) [inline tinygrad/shape/view.py:102]: Huh? This is so complex looking!

Comment (madt2709) [reply tinygrad/shape/view.py]: re read this and its complete garbage - sorry! 

Updated it again. 3rd time lucky 🤞

**Comment by geohot** [comment]
It's more readable now, but I think it should be 2 lines instead of 4, maintaining the readability and doing less useless compute.

**Comment by madt2709** [comment]
Alright it's now 2 lines instead of 4

**Comment by geohot** [comment]
Still hard to read, going to close.

**Pull request closed without merging**